[Refactor] Move classpath plugins to MockNode

williamrandolph opened the pull request:
We only use classpath plugins in tests, so we should move code handling them to our test framework. This PR introduces a new MockPluginsService in our test framework. The MockPluginsService is constructed with a list of classpath plugins that it will load by reflection. As before, these classpath plugins won't have their own classloaders.

We have to construct the pluginsService early in the `Node` class constructor so that `MockNode` methods can access it, but we have to do it after `Node` updates its `Settings` argument. Therefore, instead of passing in classpath plugins as a paramater, we are passing in a function that takes a `Settings` argument and returns a `PluginsService`. This lets us remove the classpath plugin list from the production code entirely.

Closes https://github.com/elastic/elasticsearch/issues/86635

williamrandolph committed:
Creation function instead of classpath plugin list

williamrandolph committed:
Add MockPluginsService and use it in MockNode

williamrandolph committed:
spotless

williamrandolph commented:
@elasticsearchmachine please run elasticsearch-ci/part-1

williamrandolph committed:
Remove classpathPlugins parameter from PluginsService

williamrandolph committed:
Clean up PluginsService constructor

williamrandolph committed:
Add unit tests for MockPluginsService

williamrandolph committed:
TODO on tests that use classpath plugins

williamrandolph committed:
Avoid classpath plugins or note their use in PluginsServiceTests

williamrandolph committed:
Test conventions

williamrandolph committed:
A little tidying

grcevski commented on server/src/main/java/org/elasticsearch/plugins/PluginsService.java:
```diff
@@ -79,91 +79,52 @@ record LoadedPlugin(PluginInfo info, Plugin instance, ClassLoader loader) {}
 
-        // now, find all the ones that are in plugins/
+        // load plugins
+        List<PluginInfo> pluginsList = new ArrayList<>();
```
Does it make sense for us to switch this to a Set instead of a list? It seems that PluginInfo has proper equals and hashcode, although the equals seems to also check version at the moment? It might be good to figure out if the comment on equals in PluginInfo should taken care of now as well. Should we be allowing multiple versions of equally named plugins? Same for modules, I think we shouldn't be allowing duplicates, perhaps even error out if we find some?

williamrandolph replied:
Looks like Version got added to PluginInfo a long time ago: abf9a866788b2fab855b50728670e2aa2dd42c5a

I don't see any reason in that commit for it to be there. I'll put out a quick fix for it on a different PR in case there's some sneaky dependence on it that we'd catch in tests.

We catch duplicate-named plugins and modules in our plugin-loading logic a little ways down (triggered by `loadBundles(seenBundles)`, and I added a test to catch this in https://github.com/elastic/elasticsearch/pull/86891/files#diff-099587538f96b160a9033106117540c99ffb60e00f8e4dfe8f54a2091a262943 , so I think for the moment we're good there.

williamrandolph replied:
Ah, PluginInfo's equality check was briefly commented on in PR review, but not really discussed: https://github.com/elastic/elasticsearch/pull/4378/files#r9292810

grcevski reviewed (approved):
LGTM! I left a comment around checking for duplicates on PluginInfo and perhaps looking into the comment in the PluginInfo equals method.

williamrandolph commented on server/src/test/java/org/elasticsearch/plugins/PluginsServiceTests.java:
```diff
@@ -187,11 +191,12 @@ public MultiplePublicConstructorsPlugin(final Settings settings) {
         assertThat(e, hasToString(containsString("no unique public constructor")));
     }
 
+    // TODO: avoid classpath plugins
```
Missed TODO?

williamrandolph committed:
Merge branch 'master' into refactor-plugins-service-creation-function

williamrandolph committed:
Remove Version from PluginInfo.equals

Version was added to PluginInfo equality in:

    abf9a866788b2fab855b50728670e2aa2dd42c5a

It looks like equality on name was intended before this commit, and I
don't see anything that expects plugins with the same name but different
versions to be unequal.

williamrandolph committed:
Remove done to-do

rjernst commented on server/src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -291,21 +291,26 @@ public class Node implements Closeable {
      *
      * @param initialEnvironment         the initial environment for this node, which will be added to by plugins
-     * @param classpathPlugins           the plugins to be loaded from the classpath
+     * @param pluginsServiceFunction     a function that takes a {@link Settings} object and returns a {@link PluginsService}
```
Instead of the generic "function" let's call this the ctor? ie `pluginServiceCtor`

rjernst commented on server/src/main/java/org/elasticsearch/plugins/PluginsService.java:
```diff
@@ -79,91 +79,52 @@ record LoadedPlugin(PluginInfo info, Plugin instance, ClassLoader loader) {}
-                        pluginsList.add(bundle.plugin);
-                        pluginsNames.add(bundle.plugin.getName());
-                    }
+                    plugins.stream().map(b -> b.plugin).forEach(pluginsList::add);
```
nit: the map can be `PluginBundle::plugin`

williamrandolph replied:
Doesn't seem to work because we've been directly accessing the public final `plugin` field, but it's simple to add a getter.

rjernst commented on server/src/main/java/org/elasticsearch/plugins/PluginsService.java:
```diff
@@ -79,91 +79,52 @@ record LoadedPlugin(PluginInfo info, Plugin instance, ClassLoader loader) {}
-        checkMandatoryPlugins(pluginsNames, MANDATORY_SETTING.get(settings));
+        checkMandatoryPlugins(
+            pluginsList.stream().map(PluginInfo::getName).collect(Collectors.toSet()),
+            new HashSet<>(MANDATORY_SETTING.get(settings))
```
Let's change the setting to be a Set<String> instead of List<String>?

williamrandolph replied:
It would be great if I could, but it looks like we need to add some capability to `Setting` before this would be simple.

rjernst commented on server/src/main/java/org/elasticsearch/plugins/PluginsService.java:
```diff
@@ -205,8 +166,8 @@ private static void logPluginInfo(final List<PluginInfo> pluginInfos, final Stri
-    public <T> Stream<T> map(Function<Plugin, T> function) {
-        return plugins.stream().map(LoadedPlugin::instance).map(function);
+    public final <T> Stream<T> map(Function<Plugin, T> function) {
+        return this.plugins().stream().map(LoadedPlugin::instance).map(function);
```
style nit: why add the `this`? it should be assumed for any method calls?

williamrandolph replied:
Not sure how that got in there... but I'll take it out.

rjernst commented on test/framework/src/main/java/org/elasticsearch/plugins/MockPluginsService.java:
```diff
@@ -0,0 +1,78 @@
+                "classpath plugin",
+                "NA",
+                Version.CURRENT,
+                "1.8",
```
nit: we should set this to current java version?

williamrandolph replied:
I'm going to use `JvmInfo.jvmInfo().version()`, but let me know if there's a better way.

rjernst replied:
Yeah there is a simpler way: `Runtime.version().feature()` should work.

rjernst reviewed (approved):
LGTM. I left a few nits.

williamrandolph committed:
Merge branch 'master' into refactor-plugins-service-creation-function

williamrandolph committed:
Clean up style after PR review

williamrandolph committed:
Get JVM version from Runtime

williamrandolph committed:
Merge branch 'master' into refactor-plugins-service-creation-function

williamrandolph committed:
Merge branch 'master' into refactor-plugins-service-creation-function

williamrandolph committed:
Rename method for accuracy

williamrandolph committed:
Fix merge mistake
